internal/storage/storage_integration_test.go

```diff
@@ -329,11 +332,191 @@ var _ = Describe("Storage", func() {
+
+		var (
+			claimer    *storage.Claimer
+			device     *v1alpha1.EdgeDevice
+			storageObj *v1alpha1.Storage
+		)
+
+		getCmData := func() map[string]string {
+			return map[string]string{
+				"BUCKET_HOST":   "host",
+				"BUCKET_PORT":   "443",
+				"BUCKET_NAME":   "bucket",
+				"BUCKET_REGION": "region",
+			}
+		}
+
+		getSecretData := func() map[string][]byte {
+			return map[string][]byte{
+				"AWS_ACCESS_KEY_ID":     []byte("AWS_ACCESS_KEY_ID"),
+				"AWS_SECRET_ACCESS_KEY": []byte("AWS_SECRET_ACCESS_KEY"),
+				"tls.crt":               []byte("cert"),
+			}
+		}
+
+		createS3CM := func(data map[string]string) {
+			cm := corev1.ConfigMap{
+				ObjectMeta: v1.ObjectMeta{
+					Name:      "test",
+					Namespace: "default",
+				},
+				Data: data,
+			}
+			err := k8sClient.Create(context.TODO(), &cm)
+			ExpectWithOffset(1, err).ToNot(HaveOccurred())
+		}
+
+		createS3Secret := func(data map[string][]byte) {
+			secret := corev1.Secret{
+				ObjectMeta: v1.ObjectMeta{
+					Name:      "test",
+					Namespace: "default",
+				},
+				Data: data,
+			}
+			err := k8sClient.Create(context.TODO(), &secret)
+			ExpectWithOffset(1, err).ToNot(HaveOccurred())
+		}
+
+		BeforeEach(func() {
+			claimer = storage.NewClaimer(k8sClient)
+			device = getDevice()
+			storageObj = &v1alpha1.Storage{
+				S3: &v1alpha1.S3Storage{
+					SecretName:         "test",
+					SecretNamespace:    "default",
+					ConfigMapName:      "test",
+					ConfigMapNamespace: "default",
+				},
+			}
+			device.Spec.Storage = storageObj
+		})
+
+		It("missing S3 configuration", func() {
+			// given
+			device.Spec.Storage = &v1alpha1.Storage{}
+
+			// when
+			result, err := claimer.GetExternalStorageConfig(context.TODO(), device)
+
+			// then
+			Expect(err).ToNot(BeNil())
+			Expect(result).To(BeNil())
+		})
+
+		It("missing config map", func() {
+			// given
+			device.Spec.Storage.S3.ConfigMapName = "missing"
+			createS3Secret(nil)
+
+			// when
+			result, err := claimer.GetExternalStorageConfig(context.TODO(), device)
+
+			// then
+			Expect(err).ToNot(BeNil())
+			Expect(result).To(BeNil())
+		})
+
+		It("missing secret", func() {
+			// given
+			device.Spec.Storage.S3.SecretName = "missing"
+			createS3CM(nil)
+
+			// when
+			result, err := claimer.GetExternalStorageConfig(context.TODO(), device)
+
+			// then
+			Expect(err).ToNot(BeNil())
+			Expect(result).To(BeNil())
+		})
+
+		table.DescribeTable("missing ConfigMap field",
+			func(fieldName string, optional bool) {
+				//given
+				cmData := getCmData()
+				delete(cmData, fieldName)
+				createS3CM(cmData)
+				createS3Secret(getSecretData())
+
+				// when
+				result, err := claimer.GetExternalStorageConfig(context.TODO(), device)
+
+				// then
+				if optional {
+					Expect(err).To(BeNil())
+					Expect(result).ToNot(BeNil())
+				} else {
+					Expect(err).ToNot(BeNil())
+					Expect(result).To(BeNil())
+				}
+			},
+			table.Entry("BUCKET_HOST", "BUCKET_HOST", false),
```

Comment:
If all the config map properties are compulsory, there is no need for `optional` argument. If someone adds an optional one in the future, they will have to update the test accordingly. For now let's keep it simple and please remove `optional bool` parameter and leave only assertions that apply.